### PR TITLE
Rfc/issue 185 resources refactor

### LINKED ISSUE #185: [RFC] Plugins: Transforms
## Type of Change
- [ ] New Feature Request
- [ ] Documentation / Website
- [ ] Improvement / Suggestion
- [ ] Bug
- [x] Other (please clarify below)

## Summary
Taken from a comment in https://github.com/ProjectEvergreen/greenwood/issues/17#issuecomment-523706157 in regards to a general discussion around the API and capturing the specific action item re: the `lifecycle` plugin type here.

## Details
This would essentially be enabling potentially different source file types in addition to the default markdown support, like [JavaScript](#170)  or maybe YAML?

At minimum, should probably make sure to support at least _.html_ files out of the box.

### PR DESCRIPTION
**Pull request by thescientist13**
## Related Issue
resolves #185 - a new take on an existing design as started in #450 

## Summary of Changes
1. Introduced the concept of a Resource plugin, and the `ResourceInterface` base class (f.k.a Transforms) with various lifecycles.  
1. Built in "standard" resource plugins for Greenwood for like JS, CSS, HTML / md, etc.  (note: given the tight coupling and shared logic, merged _.md_ and _.html_ back into one)
1. Refactored Koa middleware logic a bit (and more options possible using Koa Cascading middleware)
1. There is a TypeScript like example for a _.foo_ resource you can see on the home page
1. Restored / refactored Google Analytics and Polyfills plugins
1. Added tests and documentation

> _I know there is a lot of ground to cover here in this PR, so will go through a full demo / walk through during our next weekly meeting._

## Notes / Thoughts
- For GraphQL, we should now have the ability to handle `@greenwood/data` and return JS for _.gql_
  - will likely just need to expose a server's plugin API now
- For TS - _.foo_ should provide a basic POC

##  TODOs 
- [x] check to see if this new resolve logic now solves / supports #435 (for its own PR) - left a comment, will need to give Rollup a little help here
- [x] move HTML serialize teardown stuff into optimize hook in standard HTML plugin
- [x] Need at least one example of `intercept` - `importMap`?
- [x] remove example _.foo_ code from the website - you can still the same functionality in the test cases for plugins
- [x] fix unfinished comment in docs - _Note: you can always_
- [x] More issue / Trello tracking
  - do we need to add more font / icon types?
  - refactor `devServer` to use standard plugins now?
  - refactoring out parts of **StandardHtmlResource** plugin like `importMap`, `liveServer` into their own?
  - should Greenwood standard plugins be pushed to plugins array?

### CODE REVIEW

File: packages/cli/src/config/rollup.config.js [hunk 1]
```diff
@@ -37,9 +37,28 @@ function greenwoodHtmlPlugin(compilation) {
...
-        return '';
+    async load(id) {
+      const extension = path.extname(id);
+      
+      if (extension === '.html') {
+        return Promise.resolve('');
+      }
+
+      // handle custom user file extensions
+      const customResources = compilation.config.plugins.filter((plugin) => {
```
Comment (hutchgrant) [inline packages/cli/src/config/rollup.config.js:48]: Nice!

Comment (hutchgrant) [inline packages/cli/src/config/rollup.config.js:48]: Nice! Was curious how we were going to implement this in rollup.  This is exactly whats needed.

File: packages/cli/src/lib/resource-interface.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  // handle an already resolved / served resource
+  // eslint-disable-next-line no-unused-vars
+  shouldIntercept(url, headers) {
+    return false;
+  }
+
+  // eslint-disable-next-line no-unused-vars
+  async intercept(contents, headers) {
```
Comment (hutchgrant) [inline packages/cli/src/lib/resource-interface.js:42]: Interesting idea +1

File: packages/cli/src/lib/resource-interface.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    return Promise.resolve(contents);
+  }
+
+  // handle a (final) resource type post build, pre optimize, 
+  // ex: remove es shim <script>, convert .ts -> .js and update path references 
+  // this is only an _index.html_ file, BYOA (Bring Your Own AST)
+  // eslint-disable-next-line no-unused-vars
+  shouldOptimize(contents, url) {
```
Comment (hutchgrant) [inline packages/cli/src/lib/resource-interface.js:50]: Perfect for post-serialize plugin functionality

File: packages/cli/src/lifecycles/serialize.js [hunk 4]
```diff
@@ -1,44 +1,53 @@
...
             const outputPath = `${outputDir}${route}index.html`;
-            
-            // TODO allow setup / teardown (e.g. module shims, then remove module-shims)
-            let htmlModified = html;
-  
-            // TODO should really be happening via plugins or other standardize setup / teardown mechanism
-            htmlModified = htmlModified.replace(/<script src="\/node_modules\/@webcomponents\/webcomponentsjs\/webcomponents-bundle.js"><\/script>/, '');
-            htmlModified = htmlModified.replace(/<script type="importmap-shim">.*?<\/script>/s, '');
-            htmlModified = htmlModified.replace(/<script defer="" src="\/node_modules\/es-module-shims\/dist\/es-module-shims.js"><\/script>/, '');
-            htmlModified = htmlModified.replace(/<script type="module-shim"/g, '<script type="module"');
```
Comment (hutchgrant) [inline packages/cli/src/lifecycles/serialize.js:29]: :100:  agree with this refactor.

File: packages/cli/src/lifecycles/serve.js [hunk 5]
```diff
@@ -1,61 +1,109 @@
...
+    // Greenwood default standard resource plugins
+    pluginResourceStandardCss.provider(compilationCopy),
+    pluginResourceStandardFont.provider(compilationCopy),
+    pluginResourceStandardHtml.provider(compilationCopy),
+    pluginResourceStandardImage.provider(compilationCopy),
+    pluginResourceStandardJavaScript.provider(compilationCopy),
+    pluginResourceStandardJson.provider(compilationCopy),
+
+    // custom user resource plugins
+    ...compilation.config.plugins.filter((plugin) => {
```
Comment (hutchgrant) [inline packages/cli/src/lifecycles/serve.js:28]: My only concern here is we're no longer accepting pre-transform(resource) plugins and we're defaulting them all to pos transforms.  I don't have an immediate pre-default use case but for now I think this is okay.  Also another issue might be that nobody can override the default transforms(resources).

**Review by hutchgrant — APPROVED** [review]
I had got used to calling these `transforms` plugins `¯\_(ツ)_/¯`

That modification to rollup config is perfect, I was thinking people were going to have to include their own configuration with their plugins or something. The additional intercept and optimize capability is a good idea as well.

My only gripe is with no pre-transform(resource) or default override capability.  This could be addressed at a later time.

**Comment by thescientist13** [comment]
> My only gripe is with no pre-transform(resource) or default override capability. This could be addressed at a later time.

Yeah, and I think once there are some clear use cases for that, I think it will make it easier to understand what a good design for that could look like.  It _could_ be just as easy as running `shouldServe` before and after maybe?
```js
  // then handle serving urls
  app.use(async (ctx, next) => {
    // run user plugins

    // run greenwood plugins

    // run user plugins again?
  });
```

Or perhaps cascading middleware is better for this, but certainly solving in its own PR will allow us to drill down into the design and come up with something good once we have that clear use case.

----

> edit: also I'm not sure if I'm in agreement with the combined markdown and html resource. When I wrote #450 and attempted to decouple them, I recall it being slightly difficult because they both needed to make use of app/page templates.

Maybe I misunderstand your comment, but given my experience is the same as yours (decoupling meant more logic spread out in multiple places, specifically `shouldServe` type logic and thus more bugs), I also found keeping them as one actually makes our code less susceptible to errors as a result and easier to reason about (or at the very least you only have to be looking in one place).  So not sure how that makes a case for splitting them up?

**Pull request merged**